src/validation.cpp

```diff
@@ -3629,25 +3621,53 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
 
 static bool CheckBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
 {
-    /*
-     *
-     *   Header verification takes a long time due to having to compute scrypt hashes.
-     * This is a lengthy process, and so for this reason we skip it. This is acceptable
-     * since we're going to run a full validation of each and every block in the chain
-     * once the full data set is received.
-     *
-     */
-
+    // Check header only on recieving blocks
     return true;
 }
 
-static bool CheckBlockHeaderWorkHash(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
+int GetBlockHeightFromIndex(const CBlockHeader& block)
 {
-    // Check proof of work matches claimed amount
-    if (!IsBlockchainX16R() && fCheckPOW && 
-        (!CheckProofOfWork(block.GetNextMinedHash(), block.nBits, consensusParams) ||
-         !CheckProofOfWork(block.GetWorkHash(), block.nBits, consensusParams) )) 
-        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetNextMinedHash().ToString() + " proof of work failed " + block.GetWorkHash().ToString());
+    if (block.hashPrevBlock == uint256())
+    {
+        return 0;
+    }
+
+    BlockMap::iterator mi = mapBlockIndex.find(block.hashPrevBlock);
+    if (mi == mapBlockIndex.end() || mi->second == nullptr)
+    {
+        return -1;
+    }
+
+    return mi->second->nHeight + 1;
+}
+
+static bool ValidateBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
+{
+    if (fCheckPOW)
+    {
+        const int height = GetBlockHeightFromIndex(block);
+        if (height < 0)
+        {
+            return state.DoS(50, false, REJECT_INVALID, "invalid-hash", false, block.hashPrevBlock.ToString() + " block is not in index ");
+        }
+
+        if (IsPeriodX16R(consensusParams, height))
+        {
+            if (!CheckProofOfWork(block.GetHash(), block.nBits, consensusParams)) {
+                return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetHash().ToString() + " proof of work failed ");
+            }
+        }
+        else if (IsPeriodScrypt2(consensusParams, height))
+        {
+            if (block.nVersion < VERSIONBITS_TOP_BITS_SCRYPT_2) {
+                return state.DoS(50, false, REJECT_INVALID, "block-version", false, block.GetHash().ToString() + " block version failed ");
+            }
+            if (!CheckProofOfWork(block.GetWorkHash(), block.nBits, consensusParams))  {
+                return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetHash().ToString() + " proof of work failed ");
```

Comment:
It would be useful to have network period name in the message. Something like "scrypt^2 proof of work failed"

Reply:
will be added with autotests